skbio/sequence/_substitution.py

```diff
@@ -0,0 +1,518 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import numpy as np
+
+from skbio.util._decorator import experimental, classonlymethod
+from skbio.stats.distance import DissimilarityMatrix
+
+
+@experimental(as_of='0.5.10')
+class SubstitutionMatrix(DissimilarityMatrix):
+    """Matrix for scoring substitutions between characters in biological
+    sequences.
+
+    Parameters
+    ----------
+    alphabet : iterable
+        Characters that constitute the alphabet.
+    scores : 2D array-like
+        Scores of substitutions from one character (row, or axis=0) to another
+        character (column, or axis=1).
+    kwargs : dict
+        Additional arguments for the ``DissimilarityMatrix`` constructor.
+
+    See Also
+    --------
+    DissimilarityMatrix
+
+    Notes
+    -----
+    A substitution matrix (a.k.a. replacement matrix) scores the substitution
+    of each character by each other character and itself in an alphabet. The
+    score usually represents the rate of substitution over evolutionary time in
+    a biological sequence. A higher score usually indicates higher similarity
+    in chemical properties or functional roles of two molecules, therefore a
+    mutation from one to the other is easier. In sequence alignment, the score
+    can measure the likelihood that a pair of aligned characters are homologous
+    rather than by chance.
+
+    This class provides a generalized interface for substitution matrices. The
+    alphabet usually consists of individual characters, such as nucleotides or
+    amino acids, but it can be generalized to any iterable of scalars (numbers,
+    strings, etc.). Therefore, you may use this class to construct substitution
+    matrices of complicated biological units (such as codons or non-canonical
+    amino acids). The score matrix may be symmetric, as many existing matrices
+    are, or assymetric, where the score of one character substituted by another
+    is unequal to the the other way around. Only square matrices (i.e., numbers
+    of rows and columns are equal) are supported.
+
+    Multiple commonly used nucleotide and amino acid substitution matrices are
+    pre-defined and can be referred to by name. Examples include NUC.4.4 for
+    nucleotides, and variants of BLOSUM and PAM matrices for amino acids.
+
+    ``SubstitutionMatrix`` is a subclass of ``DissimilarityMatrix``. Therefore,
+    all attributes and methods of the latter also apply to the former.
+
+    Examples
+    --------
+    >>> from skbio import SubstitutionMatrix
+    >>> mat = SubstitutionMatrix('ACGT', np.array([
```

Comment:
This string literal, and the one on line 81, might make sense to not be within a variable as it is an example use as opposed to an internal library call. But at least making it possible for a user to rely on a variable would be good as it helps to future proof their code in the event skbio reconsiders e.g. the datatype used

Reply:
I don't quite understand this comment... Could you explain a bit?

Reply:
doing `>>> mat = SubstitutionMatrix(DNA_ALPHABET, ...)` or something to that effect

Reply:
Got it. I will keep this in my note. A tricky thing is that established substitution matrices have different alphabets. They diverge on whether to include degenerate codes and which degenerate codes to include. Some mechanism needs to be implemented in skbio to handle characters not found in the alphabet. For example, replacing `R` in a nucleotide sequence with `N`.

Reply:
Ah ya, that makes sense